File: workshops/2022/CS_Materials-Summer-22/openswim/33-AudioMixing/python/audio_mixing.py
from bridges import *
import json

# Goal: Implement simple mixing of the two AudioClips
def mix_clips(ac1: AudioClip, ac2: AudioClip, duration: float = -1) -> AudioClip:
    """ Return a new AudioClip of the two AudioClips mixed together
    """
    if ac1.get_sample_rate() != ac2.get_sample_rate():
        raise "two clips must have the same sampling rate"
    
    SAMPLE_RATE = ac1.get_sample_rate()
    sample_count = (SAMPLE_RATE * duration) if duration != -1 else min(ac1.get_sample_count(), ac2.get_sample_count())

    # Create a new audio clip of appropriate length
    # using the same sampling rate
    # if duration is negative, pick the shorter of the two clips

    # Mix the audio clips together and store the values in the new AudioClip
    # To mix two samples together simply take their average

    return None

def main() -> None:
    bridges = Bridges(233, "BRIDGES_USER_ID", "BRIDGES_API_KEY")

    bridges.set_title("Audio Mixing")
    bridges.set_description("Mixing two AudioClips together")

    # Load the clips
    ac_piano = AudioClip("../audio/piano.wav")
    ac_bass = AudioClip("../audio/bass.wav")

    # Apply simple mixing of the two AudioClips and visualize
    ac_mix = mix_clips(ac_piano, ac_bass)
    bridges.set_data_structure(ac_mix)
    bridges.visualize()

    #as a bonus problem. One could mix two clips with a fade in/fade out
    
if __name__ == "__main__":
    main()

File: workshops/2022/CS_Materials-Summer-22/openswim/33-AudioMixing/c++/AudioMixing.cpp
#include "AudioClip.h"
#include <iostream>
#include <fstream>
#include "Bridges.h"
#include <math.h>

using namespace std;
using namespace bridges;

/** 
 * Return a new AudioClip of the two AudioClips mixed together
*/
AudioClip mixClips(const AudioClip& ac1, const AudioClip& ac2, float duration) {
  if (ac1.getSampleRate() != ac2.getSampleRate()) {
    throw "can't mix audio clips of varying rates";
  }

  //create an audio clip configured appropriately
  int SAMPLE_RATE = ac1.getSampleRate();
  int sampleCount = SAMPLE_RATE * duration;
  
  AudioClip acMix = AudioClip(sampleCount, 1, ac1.getSampleBits(), SAMPLE_RATE);

  // Loop through the samples of both clips and use the average amplitude value
  // TODO
  
  return acMix;
}	

AudioClip mixClips(const AudioClip& ac1, const AudioClip& ac2) {
  // If no duration is given then just use the minimum song duration
  return mixClips(ac1, ac2, min((float)ac1.getSampleCount() / ac1.getSampleRate(), (float)ac2.getSampleCount()  / ac2.getSampleRate()));
}

int main() {
    Bridges bridges = Bridges(133, "BRIDGES_USER_ID", "BRIDGES_API_KEY");
    bridges.setTitle("Audio Mixing");
    bridges.setDescription("Mixing two AudioClips together");

    // Load the clips
    AudioClip acPiano = AudioClip("../audio/piano.wav");
    bridges.setDataStructure(acPiano);
    bridges.visualize();
    
    AudioClip acBass = AudioClip("../audio/bass.wav");
    bridges.setDataStructure(acBass);
    bridges.visualize();
    
    // Apply simple mixing of the two AudioClips and visualize
    AudioClip acMix = mixClips(acPiano, acBass);
    bridges.setDataStructure(acMix);
    bridges.visualize();

    // One could do a more complex mixing that fades out one clip while the other fades in
    // TODO
    
    return 0;
}
